go/ephemeral/ephemeral.go

```diff
@@ -0,0 +1,200 @@
+	if err != nil {
+		return data, err
+	}
+
+	// Sign the metadata blob with the device's long term signing key.
+	signingKey, err := g.ActiveDevice.SigningKey()
+	if err != nil {
+		return data, err
+	}
+	signedPacket, _, err := signingKey.SignToString(metadataJSON)
+
+	err = postNewDeviceEK(ctx, g, signedPacket)
+	if err != nil {
+		return data, err
+	}
+
+	return metadata, nil
+}
+
+type DeviceEKResponseElement struct {
+	MerklePayload string `json:"merkle_payload"`
+	Sig           string `json:"sig"`
+}
+
+type DeviceEKsResponse struct {
+	Results []DeviceEKResponseElement `json:"results"`
+}
+
+func GetOwnDeviceEKs(ctx context.Context, g *libkb.GlobalContext) ([]keybase1.DeviceEkMetadata, error) {
+	apiArg := libkb.APIArg{
+		Endpoint:    "user/device_eks",
+		SessionType: libkb.APISessionTypeREQUIRED,
+		NetContext:  ctx,
+		Args:        libkb.HTTPArgs{},
+	}
+	res, err := g.GetAPI().Get(apiArg)
+	if err != nil {
+		return nil, err
+	}
+
+	parsedResponse := DeviceEKsResponse{}
+	err = res.Body.UnmarshalAgain(&parsedResponse)
+	if err != nil {
+		return nil, err
+	}
+
+	// The client now needs to verify several things about these blobs its
+	// received:
+	// 1) Each is validly signed.
+	// 2) The signing key belongs to one of the current user's devices.
+	// 3) The merkle payload supplied by the server matches the hash that's
+	//    been signed over.
+	// 4) The key hasn't expired. That is, the Merkle root it was delegated
+	//    with is within one week of the current root. The server deliberately
+	//    avoids doing this filtering for us, and finding expired keys in the
+	//    results here is expected. We silently drop them.
+	currentDeviceEKs := []keybase1.DeviceEkMetadata{}
+	currentMerkleRoot, err := g.GetMerkleClient().FetchRootFromServer(ctx, libkb.EphemeralKeyMerkleFreshness)
+	for _, element := range parsedResponse.Results {
+		// Verify the sig.
+		signerKey, payload, _, err := libkb.NaclVerifyAndExtract(element.Sig)
+		if err != nil {
+			return nil, err
+		}
+
+		// Verify the signing key corresponds to a device.
+		var matchingDevice *keybase1.PublicKey
+		self, _, err := g.GetUPAKLoader().Load(libkb.NewLoadUserByUIDArg(ctx, g, g.Env.GetUID()))
+		for _, device := range self.Base.DeviceKeys {
+			if !device.KID.Equal(signerKey.GetKID()) {
+				continue
+			}
+			if device.IsRevoked {
+				return nil, fmt.Errorf("deviceEK returned for revoked device KID %s", signerKey.GetKID())
+			}
+			deviceDummyVar := device
+			matchingDevice = &deviceDummyVar
+			break
+		}
+		if matchingDevice == nil {
+			return nil, fmt.Errorf("deviceEK returned for unknown device KID %s", signerKey.GetKID())
+		}
+
+		// Decode the signed JSON.
+		var signedMetadata keybase1.DeviceEkMetadata
+		err = json.Unmarshal(payload, &signedMetadata)
+		if err != nil {
+			return nil, err
+		}
+
+		// Check the hash of the Merkle payload, confirm it matches what was signed, and parse it.
+		computedHash := sha256.Sum256([]byte(element.MerklePayload))
```

Comment:
can you convert `element.MerklePayload` to `MerkleRootPayload` type and we can just call `ShortHash()` on that?

Reply:
nvm i see we want to make sure it's good before we parse